<a href="https://colab.research.google.com/github/hhemesat/MAT-421/blob/main/ModuleA.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# 9.1 : Base-N and Binary

---



In [10]:
#First, we generate a random number to practice writing binary (base 2)
import random
list = range(1000,10000)
x = random.choice(list)
print(x)

1751


In [11]:
#We generate another random number
import random
list = range(1000,10000)
y = random.choice(list)
print(y)

2460


x is 1751 (base 10) = (1 x 2^10) + (1 x 2^9) + (0 x 2^8) + (1 x 2^7) + (1 x 2^6) + (0 x 2^5) + (1 x 2^4) +(0 x 2^3) + (1 x 2^2) + (1 x 2^1) + (1 x 2^0)
= 11011010111 (base 2) 

y is 2460 (base 10) = (1 x 2^11) +  (0 x 2^10) + (0 x 2^9) + (1 x 2^8) + ( 1 x 2^7) + (0 x 2^6) + (0 x 2^5) + (1 x 2^4) + (1 x 2^3) + (1 x 2^2) + (0 x 2^1) + (0 x 2^0)
= 100110011100 (base 2)

11011010111 + 100110011100 = 1000001110011 (base 2) = 2^12 + 2^6 + 2^5 + 2^4 + 2^1 + 2^0 
= 4211 (base 10)

In [13]:
#To verify our above result:
print(x+y)

4211


# 9.2: Floating Point Numbers

---



In [20]:
#Generate a random number
list2 = range(1,100)
z = random.choice(list2)
print(z)

20


z = 20 (base 10)

Now we convert this to IEEE754, and then find the largest number smaller than 20 and the smallest number larger than 20 

Since it is positive, the sign indicator is 0.
The largest power of 2 smaller than 20 is 4, since 2^4 is 16. Thus the characteristic is 4 + 1023 = 1027 (base 10) = 2^10 + 2^1 + 2^0 = 10000000011 (base 2)

The fraction is 20/16 - 1 = 0.25 = 1 * 1/2^2 = 0100000000000000000000000000000000000000000000000000

Thus 20 (base 10) = 0 10000000011 0100000000000000000000000000000000000000000000000000 (IEEE754)

Then the next smallest number is 0 10000000011 001111111111111111111111111111111111111111111111111

The next largest is 0 10000000011 0100000000000000000000000000000000000000000000000001 

# 9.3: Round-off Errors

---



In [14]:
#We have python print out the answer to an equation involving floating-point numbers
5.7 - 3.213

2.487

In [15]:
#However, this statement comes out false
5.7 - 3.2 == 2.487

False

In [16]:
#When using the round function, then the statement becomes true
round(5.7 - 3.213, 5) == round(2.487, 5)

True